C_hands_on_programming/ipc_methods.c

```diff
@@ -0,0 +1,358 @@
+#if 1
+void messageSend(const ipc_info *info) {
+	printf("Starting messageSend..\n"
+			"If client specified server name does not exist, this program is stuck in an infinite loop looking for a server\n"
+			);
+
+    int coid, status, size;
+    header_t header;
+    char* buffer = NULL;
+    iov_t siov[2];
+	FILE *read;
+	// open file for reading
+	read = fopen(info->filename, "rb");
+	if (read == NULL) {
+		printf("File name does not exist! Exiting the program..\n");
+		exit(EXIT_FAILURE);
+	}
+
+	printf("Looking for server..\n");
+    //Server search loop
+    while ((coid = name_open(info->argument_string, 0)) == -1)
+    {
+        sleep(1);
+    }
+    printf("Found the server connection!\n");
+
+    //Start message reading
+	fseek(read, 0, SEEK_END);
+	size = ftell(read);
+	fseek(read, 0, SEEK_SET);
+
+	header.msg_type = HEADER_IOV_TYPE;
+	header.data_size = size;
+
+	if ((buffer = (char*)malloc(size)) == NULL) {
+    	perror("malloc\n");
+    	exit(EXIT_FAILURE);
+	}
+
+
+	//read the file
+    if (fread(buffer, size, 1, read) != 1) {
+    	perror("fread\n");
+    	exit(EXIT_FAILURE);
+    }
+
+    fclose(read);
+
+	SETIOV(&siov[0], &header, sizeof(header));
+	SETIOV(&siov[1], buffer, header.data_size);
+
+	status = MsgSendvs(coid, siov, 2, NULL, 0);
+	if (status == -1) {
+		perror("MsgSendv");
+		exit(EXIT_FAILURE);
+	}
+	free(buffer);
+
+	if (MsgSendPulse(coid, -1, EOF_PULSE_CODE, 0) == -1) {
+		perror("MsgSendPulse\n");
+		exit(EXIT_FAILURE);
+	} else {
+		printf("Data is delivered. Exiting the program..\n");
+	}
+}
+#endif
+
+void messageReceive(const ipc_info *info) {
+	printf("Starting messageReceive..\n");
+	int rcvid;
+	name_attach_t *attach;
+	rbufs rbuf;
+	int eof = 0;
+
+	attach = name_attach(NULL, info->argument_string, 0);
+	while (1) {
+		rcvid = MsgReceive(attach->chid, &rbuf, sizeof(rbuf), NULL);
+		if (rcvid == -1)
+		{ //was there an error receiving msg?
+			perror("MsgReceive"); //look up errno code and print
+			exit(EXIT_FAILURE); //give up
+		} else if (rcvid == 0) {
+			if (rbuf.pulse.code == _PULSE_CODE_DISCONNECT) {
+				printf("Client %x(scoid) disconnected\n", rbuf.pulse.scoid);
+				// free up disconnected scoid
+				if (ConnectDetach(rbuf.pulse.scoid) == -1) {
+					perror("ConnectDetach\n");
+				}
+				if (eof) {
+					printf("Data is received. Exiting the program..\n");
+					break;
+				}
+			} else if (rbuf.pulse.code == EOF_PULSE_CODE) {
+				// client delivered all file content and notify the server to exit
+				printf("Client reached end of file\n");
+				eof = 1;
+
+			} else {
+				printf("Server receive a pulse code of %d and value of %x\n",
+						rbuf.pulse.code, rbuf.pulse.value.sival_int);
+			}
+			//printf("Server receive a pulse code of %d and value of %x\n", rbuf.pulse.code, rbuf.pulse.value.sival_int);
+		} else {
+			if (rbuf.type == HEADER_IOV_TYPE) {
+				char *data = NULL;
+				FILE *write;
+				// open file for writing
+				write = fopen(info->filename, "wb");
```

Comment:
I think you can improve the name of your variable. You use write to the fdes of the file to write in, but write is a verb and a POSIX function. I suggest something like file2write or something like that.

Reply:
Yeah I am aware of that, I just wanna include a MsgSendPulse from the client side for practice purpose